File: library/src/main/java/io/github/anderscheow/validator/rules/common/PastRule.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.text.DateFormat;
import java.text.ParseException;
import java.util.Date;

import io.github.anderscheow.validator.rules.BaseRule;

public class PastRule extends BaseRule {

    private DateFormat dateFormat;

    public PastRule(DateFormat dateFormat) {
        this.dateFormat = dateFormat;
    }

    public PastRule(DateFormat dateFormat, @StringRes int errorRes) {
        super(errorRes);
        this.dateFormat = dateFormat;
    }

    public PastRule(DateFormat dateFormat, @NonNull String errorMessage) {
        super(errorMessage);
        this.dateFormat = dateFormat;
    }

    @Override
    public boolean validate(String value) {
        if (value == null) {
            throw new NullPointerException();
        }

        Date parsedDate = null;
        try {
            parsedDate = dateFormat.parse(value);
        } catch (ParseException ignored) {}

        return parsedDate != null && parsedDate.before(new Date());
    }

    @NonNull
    @Override
    public String errorMessage() {
        return "Does not match past rule";
    }
}

File: library/src/main/java/io/github/anderscheow/validator/rules/common/PastRule.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.Rule
import java.text.DateFormat
import java.text.ParseException
import java.util.*

class PastRule : Rule {

    private var dateFormat: DateFormat

    constructor(dateFormat: DateFormat, @StringRes errorRes: Int) : super(errorRes) {
        this.dateFormat = dateFormat
    }

    constructor(dateFormat: DateFormat, errorMessage: String) : super(errorMessage) {
        this.dateFormat = dateFormat
    }

    override fun validate(value: String?): Boolean {
        if (value == null) {
            throw NullPointerException()
        } else {
            var parsedDate: Date? = null

            try {
                parsedDate = dateFormat.parse(value)
            } catch (ignored: ParseException) {
            }

            return parsedDate != null && parsedDate.before(Date())
        }
    }
}

fun past(dateFormat: DateFormat, @StringRes errorRes: Int): PastRule =
    PastRule(dateFormat, errorRes)

fun past(dateFormat: DateFormat, errorMessage: String): PastRule =
    PastRule(dateFormat, errorMessage)